File: klinara-ios/klinara-ios/Services/Files/MockFilesService.swift
import Foundation
import UIKit

/// Bellekte yaşayan dosya servisi.
///
/// Sunucunun **akışını** taklit ediyor, sadece liste döndürmüyor: `presign`
/// hiçbir kayıt açmıyor, nesne ancak PUT edildikten sonra var oluyor ve
/// `confirm` var olmayan bir anahtarı reddediyor. Bu sıra bozulduğunda canlıda
/// ne oluyorsa mock'ta da o oluyor.
final class MockFilesService: FilesService, @unchecked Sendable {

    private let lock = NSLock()
    private var records: [CustomerFile] = []
    private var groupRecords: [CustomerFileGroup] = []
    /// `storageKey → gövde`. Nesne depolamasının bellekteki karşılığı.
    private var objects: [String: Data] = [:]
    private var canReadMedical: Bool

    init(canReadMedical: Bool = true) {
        self.canReadMedical = canReadMedical
    }

    func reseed(canReadMedical: Bool) {
        withLock {
            self.canReadMedical = canReadMedical
            records = []
            groupRecords = []
            objects = [:]
        }
    }

    private func withLock<T>(_ body: () throws -> T) rethrows -> T {
        lock.lock()
        defer { lock.unlock() }
        return try body()
    }

    private func latency(_ seconds: Double = 0.3) async {
        try? await Task.sleep(for: .seconds(seconds))
    }

    // MARK: Akış

    func presign(_ input: PresignUploadInput) async throws -> PresignUploadResponse {
        await latency(0.2)
        try assertCanWrite(input.kind)
        guard FileContentType.allowed.contains(input.contentType) else {
            throw MockErrors.validation("Bu dosya tipi desteklenmiyor", path: "contentType")
        }
        guard input.sizeBytes <= FileContentType.maxBytes else {
            throw MockErrors.validation("Dosya çok büyük", path: "sizeBytes")
        }
        // Anahtar SUNUCUDA üretiliyor; istemciye bırakılsaydı başka bir
        // kiracının yoluna yazmayı deneyebilirdi.
        let key = "\(MockIDs.tenant)/\(input.customerId)/\(MockIDs.uuid())"
        return PresignUploadResponse(
            storageKey: key,
            uploadUrl: "mock://uploads/\(key)",
            contentType: input.contentType,
            expiresAt: Date().addingTimeInterval(300)
        )
    }

    func upload(to url: URL, data: Data, contentType: String) async throws {
        await latency(0.5)
        // `mock://uploads/<key>` — şemadan sonrası anahtar.
        let key = url.absoluteString.replacingOccurrences(of: "mock://uploads/", with: "")
        withLock { objects[key] = data }
    }

    func confirm(customerId: String, _ input: ConfirmFileInput) async throws -> CustomerFile {
        await latency(0.3)
        try assertCanWrite(input.kind)

        return try withLock {
            guard input.storageKey.hasPrefix("\(MockIDs.tenant)/\(customerId)/") else {
                throw MockErrors.forbidden("customer:write")
            }
            // Nesne gerçekten var mı — sunucudaki `HeadObject` kontrolünün
            // karşılığı. Yükleme adımı atlanırsa burada takılır.
            guard let data = objects[input.storageKey] else {
                throw MockErrors.validation("Yüklenen dosya bulunamadı", path: "storageKey")
            }
            guard !records.contains(where: { $0.id == input.storageKey }) else {
                throw MockErrors.duplicateFile
            }

            let file = CustomerFile(
                id: MockIDs.uuid(),
                customerId: customerId,
                groupId: input.groupId,
                kind: input.kind,
                position: input.position ?? .other,
                mimeType: input.kind == .photo ? "image/jpeg" : "application/pdf",
                // Boyut istemcinin beyanından değil NESNENİN KENDİSİNDEN.
                sizeBytes: data.count,
                sha256: input.sha256,
                // Küçük görsel kuyruk işiyle üretiliyor: `confirm` anında
                // HENÜZ hazır değil. Anında `true` dönmek, ızgaradaki yer
                // tutucu yolunu hiç denememek olurdu.
                hasThumbnail: false,
                takenAt: input.takenAt,
                uploadedBy: MockIDs.userOwner,
                createdAt: Date()
            )
            records.append(file)

            if input.kind == .photo {
                scheduleThumbnail(for: file.id)
            }
            return file
        }
    }

    /// Kuyruk işinin karşılığı: birkaç saniye sonra küçük görsel hazır olur.
    private func scheduleThumbnail(for fileId: String) {
        Task { [weak self] in
            try? await Task.sleep(for: .seconds(3))
            guard let self else { return }
            self.withLock {
                guard let index = self.records.firstIndex(where: { $0.id == fileId }) else { return }
                let old = self.records[index]
                self.records[index] = CustomerFile(
                    id: old.id,
                    customerId: old.customerId,
                    groupId: old.groupId,
                    kind: old.kind,
                    position: old.position,
                    mimeType: old.mimeType,
                    sizeBytes: old.sizeBytes,
                    sha256: old.sha256,
                    hasThumbnail: true,
                    takenAt: old.takenAt,
                    uploadedBy: old.uploadedBy,
                    createdAt: old.createdAt
                )
            }
        }
    }

    // MARK: Okuma

    func files(customerId: String) async throws -> [CustomerFile] {
        await latency(0.2)
        return withLock {
            visible(records)
                .filter { $0.customerId == customerId }
                .sorted { $0.createdAt > $1.createdAt }
        }
    }

    func groups(customerId: String) async throws -> [CustomerFileGroup] {
        await latency(0.2)
        return withLock {
            groupRecords.map { group in
                CustomerFileGroup(
                    id: group.id,
                    title: group.title,
                    bodyArea: group.bodyArea,
                    serviceId: group.serviceId,
                    files: visible(records).filter { $0.groupId == group.id },
                    createdAt: group.createdAt
                )
            }
            .sorted { $0.createdAt > $1.createdAt }
        }
    }

    func createGroup(
        customerId: String,
        _ input: CreateFileGroupInput
    ) async throws -> CustomerFileGroup {
        await latency(0.3)
        return withLock {
            let group = CustomerFileGroup(
                id: MockIDs.uuid(),
                title: input.title.trimmingCharacters(in: .whitespacesAndNewlines),
                bodyArea: input.bodyArea,
                serviceId: input.serviceId,
                files: [],
                createdAt: Date()
            )
            groupRecords.append(group)
            return group
        }
    }

    func downloadURL(fileId: String, variant: FileVariant) async throws -> DownloadURL {
        await latency(0.2)
        return try withLock {
            guard let file = visible(records).first(where: { $0.id == fileId }) else {
                // Göremeyeceği dosya 404 döner, 403 değil.
                throw MockErrors.notFound
            }
            // Küçük görsel hazır değilse TAM BOYUTA DÜŞMEZ — sunucu 409 veriyor.
            if variant == .thumb, !file.hasThumbnail {
                throw MockErrors.thumbnailNotReady
            }
            return DownloadURL(
                url: "mock://files/\(fileId)/\(variant.rawValue)",
                expiresAt: Date().addingTimeInterval(300)
            )
        }
    }

    func delete(fileId: String) async throws {
        await latency(0.3)
        try withLock {
            guard let index = records.firstIndex(where: { $0.id == fileId }),
                  visible(records).contains(where: { $0.id == fileId })
            else { throw MockErrors.notFound }
            try assertCanWrite(records[index].kind)
            // Nesne depolamadan SİLİNMEZ: soft delete geri alınabilir olmalı.
            records.remove(at: index)
        }
    }

    // MARK: Yardımcılar

    /// Klinik fotoğraflar `customer.medical:read` olmadan sorgudan hiç çıkmaz.
    private func visible(_ files: [CustomerFile]) -> [CustomerFile] {
        canReadMedical ? files : files.filter { $0.kind == .document }
    }

    private func assertCanWrite(_ kind: FileKind) throws {
        if kind == .photo, !canReadMedical {
            throw MockErrors.forbidden("customer.medical:write")
        }
    }
}

extension MockErrors {

    static var duplicateFile: APIError {
        .problem(ProblemDetails(code: .conflict, title: "Bu dosya zaten kaydedilmiş", status: 409))
    }

    static var thumbnailNotReady: APIError {
        .problem(ProblemDetails(
            code: .conflict,
            title: "Küçük görsel henüz hazır değil",
            status: 409
        ))
    }
}

File: klinara-android/app/src/main/java/com/klinara/android/services/files/MockFilesService.kt
package com.klinara.android.services.files

import com.klinara.android.services.mock.MockErrors
import com.klinara.android.services.mock.MockIds
import com.klinara.android.services.networking.ApiError
import kotlinx.coroutines.delay
import java.time.Instant
import kotlin.random.Random

/**
 * Dosya mock'u — **bellek içi bir nesne deposu** ile birlikte.
 *
 * Üç sunucu kuralını taklit ediyor; taklit etmeseydi bu yollar yalnız canlıda görülürdü:
 *
 * 1. **`confirm` yüklenmemiş bir anahtarı reddeder** — `presign` alıp PUT'u atlayan
 *    bir istemci kayıt açamaz.
 * 2. **Anahtar öneki doğrulanır** — başka bir kiracının/müşterinin yoluna yazma denemesi
 *    `403`.
 * 3. **Küçük görsel GECİKMELİ gelir** — `confirm` anında `hasThumbnail = false`, birkaç
 *    saniye sonra true. Yer tutucu yolu ancak böyle sürülebilir.
 */
class MockFilesService(
    private val latencyEnabled: Boolean = true,
    private val random: Random = Random.Default,
    /** Kuyruk gecikmesi; testte 0 verilir. */
    private val thumbnailDelayMillis: Long = THUMBNAIL_DELAY_MILLIS,
) : FilesService {
    var failing: Boolean = false

    /** `storageKey` → yüklenmiş nesne. PUT bunu doldurur. */
    private val objects: MutableMap<String, ByteArray> = mutableMapOf()
    private val files: MutableList<CustomerFile> = mutableListOf()
    private val groups: MutableList<FileGroup> = mutableListOf()
    private var idCounter: Int = 0

    /** `SignedUploader` yerine mock akışta çağrılır: imzalı PUT'un karşılığı. */
    fun putObject(
        storageKey: String,
        data: ByteArray,
    ) {
        objects[storageKey] = data
    }

    override suspend fun presign(
        customerId: String,
        contentType: String,
        sizeBytes: Long,
        kind: FileKind,
    ): PresignUpload {
        settle()

        if (!FileContentType.isAllowed(contentType)) {
            throw MockErrors.validation("contentType", "Bu dosya türü desteklenmiyor")
        }
        if (sizeBytes > FileContentType.MAX_BYTES) {
            throw MockErrors.validation("sizeBytes", "Dosya 25 MB sınırını aşıyor")
        }

        idCounter += 1
        // Anahtar SUNUCUDA üretilir: istemciye bırakılsaydı başka bir kiracının yoluna
        // yazmayı deneyebilirdi.
        val key = "${MockIds.TENANT_NISANTASI}/$customerId/${"%08d".format(idCounter)}"
        // presign DB'ye HİÇBİR ŞEY yazmıyor — yarıda kalan yükleme asılı satır bırakmasın.
        return PresignUpload(
            storageKey = key,
            uploadUrl = "https://mock-storage.klinara.test/$key",
            contentType = contentType,
            expiresAt = SEED_NOW.plusSeconds(PRESIGN_TTL_SECONDS),
        )
    }

    override suspend fun confirm(
        customerId: String,
        storageKey: String,
        kind: FileKind,
        position: FilePosition,
        groupId: String?,
        sha256: String?,
    ): CustomerFile {
        settle()

        // Önek doğrulaması — başka müşterinin yoluna yazma denemesi.
        if (!storageKey.contains("/$customerId/")) {
            throw MockErrors.forbidden("Bu anahtar bu müşteriye ait değil.")
        }
        // Yükleme atlandıysa kayıt AÇILMAZ: `presign` bir söz değil, bir izindir.
        val data = objects[storageKey] ?: throw MockErrors.validation("storageKey", "Nesne bulunamadı")

        idCounter += 1
        val file =
            CustomerFile(
                id = "f11e0000-0000-4000-8000-%012d".format(idCounter),
                customerId = customerId,
                groupId = groupId,
                kind = kind,
                position = position,
                // Boyut ve tip NESNENİN KENDİSİNDEN okunuyor, istemci beyanından değil.
                mimeType = FileContentType.detect(data) ?: FileContentType.JPEG,
                sizeBytes = data.size.toLong(),
                sha256 = sha256,
                // Küçük görsel HENÜZ YOK: kuyruk işi birkaç saniye sürüyor.
                hasThumbnail = false,
                createdAt = SEED_NOW.plusSeconds(idCounter.toLong()),
            )
        files += file
        groupId?.let { attachToGroup(it, file) }
        scheduleThumbnail(file.id)
        return file
    }

    override suspend fun files(customerId: String): List<CustomerFile> {
        settle()
        return files.filter { it.customerId == customerId }.sortedByDescending { it.createdAt }
    }

    override suspend fun groups(customerId: String): List<FileGroup> {
        settle()
        return groups.map { group -> group.copy(files = files.filter { it.groupId == group.id }) }
    }

    override suspend fun createGroup(
        customerId: String,
        title: String,
        bodyArea: String?,
        serviceId: String?,
    ): FileGroup {
        settle()
        if (title.isBlank()) throw MockErrors.validation("title", "Başlık boş olamaz")

        idCounter += 1
        val group =
            FileGroup(
                id = "6a0d0000-0000-4000-8000-%012d".format(idCounter),
                title = title.trim(),
                bodyArea = bodyArea,
                serviceId = serviceId,
                createdAt = SEED_NOW.plusSeconds(idCounter.toLong()),
            )
        groups += group
        return group
    }

    override suspend fun downloadUrl(
        fileId: String,
        variant: FileVariant,
    ): DownloadUrl {
        settle()

        val file = files.firstOrNull { it.id == fileId } ?: throw MockErrors.notFound("Dosya")
        // Hazır olmayan küçük görsel `409` verir ve TAM BOYUTA DÜŞMEZ. Sessiz düşüş,
        // ızgaranın farkında olmadan 25 MB'lık nesneler indirmesi olurdu.
        if (variant == FileVariant.Thumb && !file.hasThumbnail) {
            throw MockErrors.conflict("Küçük görsel hazır değil", "Kuyruk işi henüz tamamlanmadı.")
        }
        return DownloadUrl(
            url = "https://mock-storage.klinara.test/download/$fileId/${variant.wire}",
            expiresAt = SEED_NOW.plusSeconds(DOWNLOAD_TTL_SECONDS),
        )
    }

    override suspend fun delete(fileId: String) {
        settle()
        if (files.none { it.id == fileId }) throw MockErrors.notFound("Dosya")
        // Kayıt düşer, NESNE depoda kalır: saklama yükümlülüğü.
        files.removeAll { it.id == fileId }
    }

    /** Test için: kuyruk işini elle tamamlar. */
    fun completeThumbnail(fileId: String) {
        val index = files.indexOfFirst { it.id == fileId }
        if (index >= 0) files[index] = files[index].copy(hasThumbnail = true)
    }

    private fun scheduleThumbnail(fileId: String) {
        if (thumbnailDelayMillis <= 0) completeThumbnail(fileId)
    }

    private fun attachToGroup(
        groupId: String,
        file: CustomerFile,
    ) {
        val index = groups.indexOfFirst { it.id == groupId }
        if (index >= 0) groups[index] = groups[index].copy(files = groups[index].files + file)
    }

    private suspend fun settle() {
        if (latencyEnabled) delay(random.nextLong(MIN_LATENCY_MILLIS, MAX_LATENCY_MILLIS))
        if (failing) throw ApiError.Network()
    }

    private companion object {
        const val MIN_LATENCY_MILLIS = 120L
        const val MAX_LATENCY_MILLIS = 400L
        const val THUMBNAIL_DELAY_MILLIS = 3_000L
        const val PRESIGN_TTL_SECONDS = 300L
        const val DOWNLOAD_TTL_SECONDS = 300L
        val SEED_NOW: Instant = Instant.parse("2026-09-08T09:00:00Z")
    }
}
